src/lib/sedlexing.ml

```diff
@@ -117,26 +133,46 @@ let refill lexbuf =
   end
   else lexbuf.len <- lexbuf.len + n
 
+let lex_p_new_line lexbuf =
+  lexbuf.lex_curr_p <- {
+    lexbuf.lex_curr_p with
+      Lexing.
+        pos_bol = lexbuf.lex_curr_p.Lexing.pos_cnum ;
+        pos_lnum = lexbuf.lex_curr_p.Lexing.pos_lnum+1}
+
+let lex_p_incr_curr lexbuf =
+  lexbuf.lex_curr_p <- {
+    lexbuf.lex_curr_p with
+      Lexing.
+        pos_cnum = lexbuf.lex_curr_p.Lexing.pos_cnum+1 }
+
 let next lexbuf =
   let i =
     if lexbuf.pos = lexbuf.len then
       if lexbuf.finished then eof
       else (refill lexbuf; lexbuf.buf.(lexbuf.pos))
     else lexbuf.buf.(lexbuf.pos)
   in
+  incr_curr lexbuf;
+  (* '\n' = 10 *)
+  if i = 10 then new_line lexbuf;
```

Comment:
Handle `new_line` automatically on `next` is great, but why only '\n' = 10 would be treated as newline and not configurable? see also [wiki](https://en.wikipedia.org/wiki/Newline#Unicode).
Is it a option to have a mutable field for list of possible line_end characters?

Reply:
Generally speaking I'd say "\r\n" and "\n" (and only those!) should be treated as newlines by default. The various Unicode flavors out there are not particularly recommended for real world use. Some of them exist just to map chars present in historical character sets. However, I do agree that allowing the user to configure this would be reasonable. Perhaps after this gets integrated we could improve it and request a pull?

Reply:
I believe the extra work required at each character (allocating a record + a few memory reads) could slow down the lexers significantly.  See https://github.com/ocaml/ocaml/pull/1590 and https://github.com/ocaml/ocaml/pull/1585.

Also, is the code correct (incr_curr vs lex_p_incr_curr)?

Reply:
No, I think it's a topo, and should be `lex_p_incr_curr`.

Do you mean we should not update lex_curr_p in `next` but generate it when calling `lexing_positions`, `mark` and `backtrace`? (just like what #58 do)

What about new_line detection? I think it should be in `next` to get most accurate column info.

Reply:
I'm just saying that automatically updating the field has a cost, and some application are not interesting in tracking line numbers and might not want to pay the extra cost.

One solution is to do like in #1590, i.e. update positions only if they are not `== dummy_pos`.

Reply:
@pmetzger Yes we are doing so.

I also don't know how much the cost is. If you meet some performance issue or would be willing to do some performance test after checked in, we could try to do some work around it.

Reply:
To give an idea of the cost, for ocamllex:

https://github.com/ocaml/ocaml/pull/1585/commits/bbbfefc1214dd490650538b9d5558e0b67545342

In that benchmark, lexing throughput moved from 87 Mb/s to 134 Mb/s.

Reply:
> In that benchmark, lexing throughput moved from 87 Mb/s to 134 Mb/s.

That seems like a shame. When I'm writing lexers in C I never have overhead like that for something small like tracking line numbers. Is it possible that the compiler might be able to optimize this case better with careful coding, say by doing something intelligent about the record allocation?

Reply:
Not possible with current workflow I think. (correct me if I'm wrong)
When we change the `lex_curr_p`, it is possible referenced also by `lex_mark_p`, so compiler would have no idea in compile time if it could reuse the record and have to create a new one and wait the old one to GC.

My concern is if the cost is mainly on memory allocation and GC. If so, could we make `lex_curr_p` and `lex_mark_p` plain in lexbuf to avoid them?

Reply:
> If so, could we make lex_curr_p and lex_mark_p plain in lexbuf to avoid them?

I believe this would probably help a lot.